File: block.go
package main

import (
	"bytes"
	"crypto/sha256"
	"encoding/gob"
	"fmt"
	"time"
)

type Block struct {
	Timestamp     int64          // Create timestamp
	Transactions  []*Transaction // Transactions
	PrevBlockHash []byte         // Hash value of previous Block
	Hash          []byte         // Hash value of this Block
	Nonce         int            // Nonce used to generate Hash
}

// NewBlock creates new Block
func NewBlock(transactions []*Transaction, prevBlockHash []byte) *Block {
	block := &Block{
		Timestamp:     time.Now().Unix(),
		Transactions:  transactions,
		PrevBlockHash: prevBlockHash,
		Hash:          []byte{},
	}
	pow := NewProofOfWork(block)
	nonce, hash := pow.Run()

	block.Hash = hash
	block.Nonce = nonce
	return block
}

func (b *Block) Serialize() []byte {
	var result bytes.Buffer

	encoder := gob.NewEncoder(&result)

	if err := encoder.Encode(b); err != nil {
		_ = fmt.Errorf("Block serialization failed\n")
	}
	return result.Bytes()
}

func DeserializeBlock(d []byte) *Block {
	var block Block

	decoder := gob.NewDecoder(bytes.NewReader(d))

	if err := decoder.Decode(&block); err != nil {
		_ = fmt.Errorf("Block deserialization failed\n")
	}

	return &block
}

func (b *Block) HashTransactions() []byte {
	var txHashes [][]byte
	var txHash [32]byte

	for _, tx := range b.Transactions {
		txHashes = append(txHashes, tx.ID)
	}
	txHash = sha256.Sum256(bytes.Join(txHashes, []byte{}))
	return txHash[:]
}

File: src/block.rs
use std::time::{Duration, SystemTime};

use crate::proofofwork::ProofOfWork;

pub struct Block {
    timestamp: Duration,
    data: Vec<u8>,
    prev_block_hash: [u8; 32],
    hash: [u8; 32],
    nonce: u64,
}

impl Block {
    pub fn new(data: &str, prev_block_hash: &[u8; 32]) -> Self {
        let timestamp = SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap();
        let data = Vec::from(data.as_bytes());

        let mut block = Block {
            timestamp,
            data,
            prev_block_hash: *prev_block_hash,
            hash: [0; 32],
            nonce: 0,
        };

        let pow = ProofOfWork::new(&block);
        let (nonce, hash) = pow.run();
        block.hash = hash;
        block.nonce = nonce;
        block
    }

    pub fn timestamp(&self) -> &Duration {
        &self.timestamp
    }

    pub fn data(&self) -> &Vec<u8> {
        &self.data
    }

    pub fn prev_block_hash(&self) -> &[u8; 32] {
        &self.prev_block_hash
    }

    pub fn hash(&self) -> &[u8; 32] {
        &self.hash
    }

    pub fn nonce(&self) -> u64 {
        self.nonce
    }
}
